Add "playlist" parameter, refactor yearsFilter

flomnes opened the pull request:
New parameter playlist to be used with xpansion, see [this PR](https://github.com/AntaresSimulatorTeam/antares-xpansion/pull/50)

flomnes committed:
Convert yearsFilter from bool* to std::vector<bool>

flomnes committed:
Add "playlist" parameter

flomnes committed:
Restore existing behavior

flomnes committed:
Restore existing behavior (bis)

flomnes committed:
Add warning against deprecated options

flomnes committed:
Add version number in parameter write

To allow for a different treatment depending on study version number.

guilpier-code commented on src/libs/antares/study/parameters.cpp:
```diff
@@ -612,8 +610,14 @@ static bool SGDIntLoadFamily_N(Parameters& d, const String& key, const String& v
 
 static bool SGDIntLoadFamily_P(Parameters& d, const String& key, const String& value, uint)
 {
+    const auto warning = [&logs, &key]() {
```
Instead of "Consider using playlist instead.", rather "Please save your study to upgrade it."

guilpier-code replied:
This code does not compile on my machine : 
**parameters.cpp(613,38): error C3480: 'Antares::logs': a lambda capture must be a variable from an enclosing function scope**

jmkerloch replied:
@guilpier can you indicate MSVC version used ?

guilpier-code replied:
This problem was solved for me in latest commit

flomnes committed:
Fix compilation

flomnes committed:
Change warning message

flomnes committed:
Use logs.info() instead of logs.warning()

warning() stops the execution of some studies, breaks the CI

flomnes committed:
Add write of weights

flomnes committed:
Fix weight write

flomnes committed:
Rename "playlist" to "active_years"

flomnes committed:
Add section [playlist] only if needed

flomnes committed:
Fix C++

flomnes commented on src/libs/antares/study/parameters.cpp:
```diff
@@ -1045,26 +1069,7 @@ bool Parameters::loadFromINI(const IniFile& ini, uint version, const StudyLoadOp
     // forcing value
     if (options.nbYears != 0)
     {
-        if (options.nbYears > nbYears)
-        {
-            if (yearsFilter)
-            {
-                // The variable `yearsFilter` must be enlarged
-                bool* newset = new bool[options.nbYears];
-                for (uint i = 0; i != nbYears; ++i)
-                    newset[i] = yearsFilter[i];
-                for (uint i = nbYears; i < options.nbYears; ++i)
-                    newset[i] = false;
-                delete[] yearsFilter;
-                yearsFilter = newset;
-            }
-            else
-            {
-                yearsFilter = new bool[options.nbYears];
-                for (uint i = 0; i < options.nbYears; ++i)
-                    yearsFilter[i] = true;
-            }
-        }
```
Behavior is different when `options.nbYears <= nbYears`

flomnes replied:
behavior is the same, my mistake

flomnes committed:
Add read for new format weights

flomnes committed:
Fix warning related to comparison

flomnes committed:
Rename active_years -> playlist_active_years

flomnes committed:
Rename weights[y] -> playlist_weight_year_y

entries with [] are not supported by the .ini parser, they are
mistaken for section names

flomnes committed:
Fix section name

flomnes committed:
Formatting

flomnes committed:
Add read & write support for "all_but" in playlist_active_years

flomnes committed:
Revert "Add read & write support for "all_but" in playlist_active_years"

This reverts commit 5b1227aeb4d88d0718cf3f3ccaac0ce783ff5ad3.

flomnes committed:
Merge remote-tracking branch 'origin/develop' into feature/mc-playlist

flomnes committed:
Fix build

flomnes committed:
Fix Boost import by Cmake, add boost::filesystem

flomnes committed:
Make weights public in class Parameters

flomnes committed:
Fix end-to-end-simple-study's label

TESTLABEL was also a label, probably a copy/paste error

flomnes committed:
Add playlist read tests

flomnes committed:
Run "playlist" C/I

flomnes committed:
Fix CI

For Windows (string type) and CentOS 7 (boost::filesystem has no class ofstream).

flomnes committed:
Use BOOST_CHECK and BOOST_CHECK_EQUAL instead of BOOST_TEST

flomnes committed:
Use Catch2 and std::filesystem for playlist unit tests

Boost has several problems :
- An old incompatible version (1.53) is used on CentOS 7
- Paths are not encoded as std::strings in Windows

This breaks the CI. So let's try Catch2 !

flomnes committed:
Add support for std::experimental::filesystem

For GCC < 8

flomnes committed:
Fix fs linking for gcc 7, 8

flomnes committed:
Fix build on Windows

flomnes committed:
simple-study : use std::fabs instead of tt::tolerance

tt::tolerance is not supported on CentOS 7 (Boost 1.53)

flomnes committed:
Use BOOST_CHECK instead of BOOST_TEST

for compatibility reasons

flomnes committed:
Add write tests vor v810

flomnes committed:
Rename test "playlist" -> "study-rw" and add documentation

jmkerloch commented on src/libs/antares/study/parameters.cpp:
```diff
@@ -650,8 +658,26 @@ static bool SGDIntLoadFamily_P(Parameters& d, const String& key, const String& v
     }
 
+    /* Format added in Antares 8.1 */
+    if (key == "playlist_active_years")
+    {
+        std::list<int> in_playlist;
+        value.split(in_playlist, ",");
+        // Check indices
+        for (int y : in_playlist)
+            if ((uint)y >= d.nbYears || y < 0)
+                return false;
+        // All indices good, change d.yearsFilter
+        for (uint y = 0; y < d.nbYears; ++y)
+            d.yearsFilter[y] = false;
+        for (int y : in_playlist)
+            d.yearsFilter[y] = true;
+        return true;
```
So if we want to activate all but deactivate a few years, we need to indicate all the activated years ?

flomnes replied:
right

jmkerloch commented on src/libs/antares/study/parameters.cpp:
```diff
@@ -697,6 +723,35 @@ static bool SGDIntLoadFamily_P(Parameters& d, const String& key, const String& v
     }
 
+    /* Format added in Antares 8.1 */
+    if (key.startsWith("playlist_weight_year_"))
+    {
+        String key_cpy(key);
+        uint year_in;
+        float weight_in;
+        key_cpy.replace("playlist_weight_year_", "");
+        if (!key_cpy.to<uint>(year_in))
+        {
+            logs.error() << "parameters: invalid MC year index. Got '" << key << "'";
+            return false;
+        }
+        if (!value.to<float>(weight_in))
+        {
+            logs.error() << "parameters: invalid MC year weight. Got '" << value << "'";
+            return false;
+        }
+
+        if (weight_in < 0.f)
+        {
+            logs.warning() << "parameters: invalid MC year weight.Got '" << weight_in
+                           << "'. Value not used";
+            weight_in = 1.f;
+        }
+
+        d.setYearWeight(year_in, weight_in);
+        return true;
```
Could we use a kind of a map ?
`{1: 2.3 , 2 : 4.5}`

flomnes replied:
we could, but it's harder to parse and possibly less human-readable.

jmkerloch commented on src/libs/antares/study/parameters.cpp:
```diff
@@ -1045,26 +1100,7 @@ bool Parameters::loadFromINI(const IniFile& ini, uint version, const StudyLoadOp
-            }
-            else
-            {
-                yearsFilter = new bool[options.nbYears];
-                for (uint i = 0; i < options.nbYears; ++i)
-                    yearsFilter[i] = true;
```
If there was no yearsFilter we init filter with true.

With new implementation resize is done with false value.

To be discussed.

flomnes replied:
done

jmkerloch commented on src/libs/antares/study/parameters.cpp:
```diff
@@ -1697,49 +1725,88 @@ void Parameters::saveToINI(IniFile& ini) const
+        if (version <= version800)
         {
-            if (yearsFilter[i])
+            // Playlist section must be added if at least one year is disable or one MC year weight
+            // sum is not equal to nbYears
```
report comment with correction for MC years weight check

flomnes replied:
I'm not sure I understand, can you clarify ?

jmkerloch commented on src/tests/end-to-end/CMakeLists.txt:
```diff
@@ -18,20 +34,46 @@ target_link_libraries(tests-simple-study
 						libantares-solver-ts-generator
 						libmodel_antares
-						Boost::unit_test_framework
+                        Catch2::Catch2
```
BOOST still used for simple-study.cpp

jmkerloch commented on src/tests/end-to-end/playlist.cpp:
```diff
@@ -0,0 +1,256 @@
+playlist_year += 1)");
+
+    StudyLoadOptions loadOpt;
+    loadOpt.nbYears = 5;
```
We could add a test with a `loadOpt.nbyears` different than `nbyears `in `generaldata.ini`

flomnes replied:
what is the expected behavior ?

flomnes replied:
```
/home/florian/Antares_Simulator/src/tests/end-to-end/playlist.cpp:110: FAILED:
  REQUIRE( p.nbYears == 5 )
with expansion:
  3 == 5
```

flomnes replied:
Apparently, `loadOpt.nbYears` overrides `generaldata.ini`'s `nbyears`.

flomnes replied:
I will leave `loadOpt.nbYears=5` for now. I do not know if we want to support other values than the one provided in `generaldata.ini`.

jmkerloch reviewed (changes requested):
Some questions and remarks before merge

flomnes committed:
PR remarks

flomnes committed:
Apply clang-format

flomnes committed:
Add boost version

flomnes reviewed (commented):
I'll keep the current syntax as is for now
```ini
playlist_weight_year_1 = 2.2
playlist_weight_year_3 = 1.1
```
@jmkerloch suggests
```ini
playlist_weights = {1 : 2.2, 3 : 1.1}
```

flomnes committed:
Merge remote-tracking branch 'origin/develop' into feature/mc-playlist

flomnes commented:
After some discussions with the development team, the decision was made not to include this change to the next version of Antares Simulator.

Main reason : compatibility issues. Antares API should make these changes less useful in the future.
